examples/dpi/dpi.go

```diff
@@ -0,0 +1,374 @@
+	flag.Parse()
+
+	// Initialize NFF-Go library
+	config := flow.Config{}
+	CheckFatal(flow.SystemInit(&config))
+
+	var err error
+	rules, err = getRulesFromFile(rulesfile)
+	CheckFatal(err)
+
+	if !useHS {
+		packetFilter = filterByRegexp
+	} else {
+		packetFilter = filterByHS
+		setupHyperscan(rules)
+	}
+
+	// Receive packets from given PCAP file
+	inputFlow := flow.SetReceiverFile(infile, int32(nreads))
+
+	// Split packets into flows by hash of five-tuple
+	// Packets without five-tuple are put in last flow and will be dropped
+	outputFlows, err := flow.SetSplitter(inputFlow, splitBy5Tuple, totalNumFlows, nil)
+	CheckFatal(err)
+
+	// Drop last flow
+	CheckFatal(flow.SetStopper(outputFlows[totalNumFlows-1]))
+
+	for i := uint(0); i < numFlows; i++ {
+		lc := localCounters{handlerId: i}
+		CheckFatal(flow.SetHandlerDrop(outputFlows[i], packetFilter, lc))
+	}
+
+	outFlow, err := flow.SetMerger(outputFlows[:numFlows]...)
+	CheckFatal(err)
+
+	CheckFatal(flow.SetSenderFile(outFlow, outfile))
+
+	go func() {
+		CheckFatal(flow.SystemStart())
+	}()
+
+	// Finish by timeout, as cannot verify if file reading finished
+	time.Sleep(timeout)
+
+	// Compose info about all handlers
+	var read uint64
+	var allowed uint64
+	var blocked uint64
+	fmt.Println("\nHandler statistics")
+	for i := uint(0); i < numFlows; i++ {
+		fmt.Printf("Handler %d processed %d packets (allowed=%d, blocked by signature=%d)\n",
+			i, readPktsCount[i], allowedPktsCount[i], blockedPktsCount[i])
+		read += readPktsCount[i]
+		allowed += allowedPktsCount[i]
+		blocked += blockedPktsCount[i]
+	}
+	fmt.Println("Total:")
+	fmt.Println("read =", read)
+	fmt.Println("allowed =", allowed)
+	fmt.Println("dropped (read - allowed) =", read-allowed)
+	if !useHS {
+		fmt.Println("blocked =", blocked)
+	}
+
+	if useHS {
+		cleanupHyperscan()
+	}
+}
+
+type localCounters struct {
+	handlerId         uint
+	allowedCounterPtr *uint64
+	readCounterPtr    *uint64
+	blockedCounterPtr *uint64
+}
+
+// Create new counters for new handler
+func (lc localCounters) Copy() interface{} {
+	var newlc localCounters
+	// Clones has the same id
+	id := lc.handlerId
+	newlc.handlerId = id
+	newlc.allowedCounterPtr = &allowedPktsCount[id]
+	newlc.readCounterPtr = &readPktsCount[id]
+	newlc.blockedCounterPtr = &blockedPktsCount[id]
+	return newlc
+}
+
+func (lc localCounters) Delete() {
+}
+
+func filterByRegexp(pkt *packet.Packet, context flow.UserContext) bool {
+	cnt := context.(localCounters)
+	numRead := cnt.readCounterPtr
+	numAllowed := cnt.allowedCounterPtr
+	numBlocked := cnt.blockedCounterPtr
+
+	atomic.AddUint64(numRead, 1)
+	data, err := extractData(pkt)
+	if err != nil {
+		fmt.Println("WARNING:", err, ", drop packet")
+		return false
+	}
+
+	accept := false
+	for _, rule := range rules {
+		result := rule.Re.Match(data)
+		if !result {
+			continue
+		}
+		if rule.Allow {
+			accept = true
+		} else {
+			accept = false
+			atomic.AddUint64(numBlocked, 1)
```

Comment:
So if 1 packet is disallowed by multiple rules we will increment numBlocked multiple times?
Looks like we can have numBlocked > numRead, which is strange